### PR TITLE
[python] Update chunk-sizing estimator for chunked matrix write

### PR DESCRIPTION
**Pull request by johnkerl**
**Issue and/or context:** GitHub issue #1783 and Shortcut story [sc-28843].

### CODE REVIEW

**Comment by johnkerl** [comment]
@gspowley is on PTO this week

@nguyenv or @aaronwolen do you have time to take a look?

File: apis/python/src/tiledbsoma/io/ingest.py [hunk 1]
```diff
@@ -1921,20 +2097,75 @@ def _coo_to_table(
...
+        while chunk_coo.nnz > tiledb_create_options.goal_chunk_nnz:
+            num_tries += 1
+            # The logger we use doesn't have a TRACE level. If it did, we'd use it here.
+            # logging.logger.trace(
+            #    f"Adapt: {num_tries}/{max_tries} {chunk_coo.nnz}/{tiledb_create_options.goal_chunk_nnz}"
+            # )
+            if num_tries > max_tries:
+                raise SOMAError(
+                    f"Unable to accommodate goal_chunk_nnz {goal_chunk_nnz}. "
+                    + "This may be reduced in TileDBCreateOptions."
```
Comment (nguyenv) [inline apis/python/src/tiledbsoma/io/ingest.py:2154]: Nit: the plus sign isn't necessary.

File: apis/python/src/tiledbsoma/io/ingest.py [hunk 2]
```diff
@@ -1921,20 +2097,75 @@ def _coo_to_table(
...
+                )
+
+            ratio = chunk_coo.nnz / tiledb_create_options.goal_chunk_nnz
+            chunk_size = int(math.floor(0.9 * (i2 - i) / ratio))
+            if chunk_size < 1:
+                raise SOMAError(
+                    f"Unable to accommodate a single row at goal_chunk_nnz {goal_chunk_nnz}. "
+                    + "This may be reduced in TileDBCreateOptions."
```
Comment (nguyenv) [inline apis/python/src/tiledbsoma/io/ingest.py:2162]: Nit: the plus sign isn't necessary.

File: apis/python/src/tiledbsoma/io/ingest.py [hunk 3]
```diff
@@ -1776,46 +1780,217 @@ def _find_sparse_chunk_size(
...
+    Then in the non-backed case we simply check each row's nnz value. But for
+    the backed case, that absolutely tanks performance.
+
+    Another option is getting a sample of nnz values and using them as an
+    estimator for nnz values of other rows. This often works, but there's enough
+    diversity in the row nnzs that this estimator can result in us sending
+    too-big chunks to remote services.
+
+    It turns out, though, as a peculiarity of AnnData backed matrices, that
+    while it's prohibitively expensive to ask the 1000 questions
```
Comment (nguyenv) [inline apis/python/src/tiledbsoma/io/ingest.py:1893]: Nit: 10,000?

**Pull request merged**